flink-cdc-common/src/main/java/org/apache/flink/cdc/common/data/binary/BinaryArrayData.java

```diff
@@ -0,0 +1,574 @@
+
+            if (value == null) {
+                setNullAt(pos);
+                // keep the offset for future update
+                BinarySegmentUtils.setLong(segments, fieldOffset, ((long) cursor) << 32);
+            } else {
+
+                byte[] bytes = value.toUnscaledBytes();
+                assert (bytes.length <= 16);
+
+                // Write the bytes to the variable length portion.
+                BinarySegmentUtils.copyFromBytes(segments, offset + cursor, bytes, 0, bytes.length);
+                setLong(pos, ((long) cursor << 32) | ((long) bytes.length));
+            }
+        }
+    }
+
+    public void setTimestamp(int pos, TimestampData value, int precision) {
+        assertIndexIsValid(pos);
+
+        if (TimestampData.isCompact(precision)) {
+            setLong(pos, value.getMillisecond());
+        } else {
+            int fieldOffset = getElementOffset(pos, 8);
+            int cursor = (int) (BinarySegmentUtils.getLong(segments, fieldOffset) >>> 32);
+            assert cursor > 0 : "invalid cursor " + cursor;
+
+            if (value == null) {
+                setNullAt(pos);
+                // zero-out the bytes
+                BinarySegmentUtils.setLong(segments, offset + cursor, 0L);
+                // keep the offset for future update
+                BinarySegmentUtils.setLong(segments, fieldOffset, ((long) cursor) << 32);
+            } else {
+                // write millisecond to the variable length portion.
+                BinarySegmentUtils.setLong(segments, offset + cursor, value.getMillisecond());
+                // write nanoOfMillisecond to the fixed-length portion.
+                setLong(pos, ((long) cursor << 32) | (long) value.getNanoOfMillisecond());
+            }
+        }
+    }
+
+    public boolean anyNull() {
+        for (int i = offset + 4; i < elementOffset; i += 4) {
+            if (BinarySegmentUtils.getInt(segments, i) != 0) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void checkNoNull() {
+        if (anyNull()) {
+            throw new RuntimeException("Primitive array must not contain a null value.");
+        }
+    }
+
+    @Override
+    public boolean[] toBooleanArray() {
+        checkNoNull();
+        boolean[] values = new boolean[size];
+        BinarySegmentUtils.copyToUnsafe(
+                segments, elementOffset, values, BOOLEAN_ARRAY_OFFSET, size);
+        return values;
+    }
+
+    @Override
+    public byte[] toByteArray() {
+        checkNoNull();
+        byte[] values = new byte[size];
+        BinarySegmentUtils.copyToUnsafe(
+                segments, elementOffset, values, BYTE_ARRAY_BASE_OFFSET, size);
+        return values;
+    }
+
+    @Override
+    public short[] toShortArray() {
+        checkNoNull();
+        short[] values = new short[size];
+        BinarySegmentUtils.copyToUnsafe(
+                segments, elementOffset, values, SHORT_ARRAY_OFFSET, size * 2);
```

Comment:
Why not get SHORT_ARRAY_OFFSET when used rather than when initialization?

Reply:
are you suggesting removal of static variable SHORT_ARRAY_OFFSET and computing it directly instead?

Reply:
I just wonder why design like it.

Reply:
Its a port of the BinaryArrayData from flink source as mentioned [here](https://github.com/apache/flink-cdc/pull/3434#issuecomment-2518863198)